### PR TITLE
Extended loss

### PR DESCRIPTION
**Pull request by jj-globant**
This is the plan described on #271  where ONMTDataset is extended so it can handle an extra field, dw (for datum-weight) that will enable to use a value that represents how authoritative a single datum is, to weight the calculation of the loss and affect less the ones which i want to use for the sake of completeness but that shouldn't be as important as the most authoritative ones.

Also, a custom Loss object was created with it's own criterium, following closely the NLLLoss but with datum-weights.

Train objects were adapted to enable using the extra argument, and the `opt` objects include an argument to select the Loss function, defaulting to the standard one.

This PR supersedes #278

### CODE REVIEW

File: onmt/Loss.py [hunk 1]
```diff
@@ -103,6 +104,106 @@ def compute_loss(self, batch, output, target, **kwargs):
...
+        Args:
+            loss(Tensor): the loss computed by the loss criterion.
+            scores(Tensor): a sequence of predict output with scores.
+        """
+        pred = scores.max(1)[1]
+        non_padding = target.ne(self.padding_idx)
+        num_correct = pred.eq(target) \
+                          .masked_select(non_padding) \
+                          .sum()
+        return onmt.Statistics(loss[0], non_padding.sum(), num_correct)
```
Comment (JianyuZhan) [inline onmt/Loss.py]: If the  `NMTLossComputeDatumWeighted` needs a new `stats()` method,  just override the base class version in `NMTLossComputeDatumWeighted`,  I think we don't need to put it in  the `NMTLossCompute` itself.

File: onmt/Loss.py [hunk 2]
```diff
@@ -103,6 +104,106 @@ def compute_loss(self, batch, output, target, **kwargs):
...
+
+
+class NMTLossComputeDatumWeighted(LossComputeBase):
+    """
+    Standard NMT Loss Computation.
+    """
```
Comment (JianyuZhan) [inline onmt/Loss.py]: Will we make this standard one?

File: onmt/Loss.py [hunk 3]
```diff
@@ -103,6 +104,106 @@ def compute_loss(self, batch, output, target, **kwargs):
...
+    def __init__(self, generator, tgt_vocab):
+        super(NMTLossComputeDatumWeighted, self).__init__(generator, tgt_vocab)
+        self.copy_attn = False
+        weight = torch.ones(len(tgt_vocab))
+        weight[self.padding_idx] = 0
+        self.criterion = DatumWeightedNLLCriterion(weight, size_average=False)
+
+    def compute_loss(self, batch, output, target, **kwargs):
```
Comment (JianyuZhan) [inline onmt/Loss.py]: For this concrete subclass, we don't want a vague `kwargs` here,  just list all args here it needs, not default `None` needed,  because if we use this `NMTLossComputeDatumWeighted`,  we already make sure these args are satisfied. 
You can refer to `onmt/CopyGenerator.py::CopyGeneratorLossCompute`.

File: onmt/Loss.py [hunk 4]
```diff
@@ -103,6 +104,106 @@ def compute_loss(self, batch, output, target, **kwargs):
...
+        super(NLLLoss, self).__init__(weight, size_average)
+        self.ignore_index = ignore_index
+        self.datum_average = datum_average
+
+    @staticmethod
+    def _assert_no_grad(variable):
+        assert not variable.requires_grad, \
+            "nn criterions don't compute the gradient w.r.t. targets - " \
+            "please mark these variables as volatile or not requiring" \
+            "gradients"
```
Comment (JianyuZhan) [inline onmt/Loss.py]: A private method makes no sense to make it staticmethod?

File: onmt/Loss.py [hunk 5]
```diff
@@ -113,12 +214,15 @@ def make_gen_state(output, batch, attns, range_, copy_attn=None):
...
+    use_dw = hasattr(batch, "dw")
     return {"output": output,
             "target": batch.tgt[range_[0] + 1: range_[1]],
             "copy_attn": attns.get("copy"),
             "align": None if not copy_attn
             else batch.alignment[range_[0] + 1: range_[1]],
-            "coverage": attns.get("coverage")}
+            "coverage": attns.get("coverage"),
+            "dw": None if not use_dw
+            else batch.dw[range_[0] + 1: range_[1]]}
```
Comment (JianyuZhan) [inline onmt/Loss.py]: Maybe later we could make this subclass specific to get rid of these `if...else...` ?

File: onmt/Loss.py (hunk 1 above)
Comment (JianyuZhan) [reply onmt/Loss.py]: Besides,  to be consistent with existing `*LossCompute` classes,  it would be better call it `DatumWeightedNMTLossCompute`

File: onmt/Loss.py (hunk 5 above)
Comment (jj-globant) [reply onmt/Loss.py]: definitely... should opts be propagated to avoid including `copy_attn`, `align` and `dw`?

File: onmt/Loss.py (hunk 1 above)
Comment (jj-globant) [reply onmt/Loss.py]: I'm not sure about the `stats()` really... it was more of an open question and placeholder. I think the stats are what they should be, but maybe someone had an oppinion about weighting the `num_correct` or something. What's your take on that?

Comment (JianyuZhan) [reply onmt/Loss.py]: For the `NMTLossCompute`, if we use the current `ONMTDataset`, the `stats()` in the base class would do,  no need to re-implement it in `NMTLossCompute`; if we use the weighted `ONMTDataset`, I think we should use`DatumWeightedNMTLossCompute`, not the `NMTLossCompute`, right?(If someone specifies -train_dw and -valid_dw, but still use the standard `NMTLossCompute`, it is weird, and I think we should add constraint here to not let user do so).  Again, no need to touch  `NMTLossCompute`.  So I am wondering why you put it in `NMTLossCompute`.

Then comes to question whether we need a `weighted` version of `stats()` for  the `DatumWeightedNMTLossCompute`?  We can think a little bit. The purpose of incorporating a `weighted` version of source data is due to the different degree of trustworthiness of each data example, I would interpret this as sort of data augmentation. But for the `num_correct` in `stats()`, it just tells how many words we predict correctly compared to the baseline `target`. In this sense, 1 is just 1,  I think we don't need to `weight` this figure here.

File: onmt/Loss.py (hunk 5 above)
Comment (JianyuZhan) [reply onmt/Loss.py]: No.  Keep the `*LossCompute` class getting what they really need. We would like to make this like a library that can be used directly from pytorch someday, so it should not care about whether these settings are from command line or other approaches.

File: onmt/Loss.py (hunk 1 above)
Comment (jj-globant) [reply onmt/Loss.py]: you are right... there is no need so far to affect `stats()`, and it was in the wrong class anyway! just removed it on the latest commit.

File: onmt/Loss.py (hunk 5 above)
Comment (jj-globant) [reply onmt/Loss.py]: i'll leave it be for now then!

**Comment by JianyuZhan** [comment]
Hi, @jj-globant, the LossCompute interface is now more neat, are you willing to update your work to make it merged? If you want to, I can provide help should you need any. Thanks!

**Comment by jj-globant** [comment]
@JianyuZhan I will do! Might not be too fast, but gladly!

**Comment by JianyuZhan** [comment]
@jj-globant , great! Feel free to ping me if you need any help!

**Comment by jj-globant** [comment]
@vince62s I'm planning to, but my timing is not the best, so i'm open to go at it with collaborators!
